File: clothing-erp-project/frontend/lib/models/product.dart
class ProductSku {
  final String? id;
  final String? colorName;
  final String? sizeName;
  final String? barcode;
  final double? retailPrice;
  final double? wholesalePrice;
  final double? purchasePrice;
  final int? stockQty;

  ProductSku({
    this.id,
    this.colorName,
    this.sizeName,
    this.barcode,
    this.retailPrice,
    this.wholesalePrice,
    this.purchasePrice,
    this.stockQty,
  });

  factory ProductSku.fromJson(Map<String, dynamic> json) {
    return ProductSku(
      id: json['id'] as String?,
      colorName: json['colorName'] as String?,
      sizeName: json['sizeName'] as String?,
      barcode: json['barcode'] as String?,
      retailPrice: (json['retailPrice'] as num?)?.toDouble(),
      wholesalePrice: (json['wholesalePrice'] as num?)?.toDouble(),
      purchasePrice: (json['purchasePrice'] as num?)?.toDouble(),
      stockQty: json['stockQty'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'colorName': colorName,
      'sizeName': sizeName,
      'barcode': barcode,
      'retailPrice': retailPrice,
      'wholesalePrice': wholesalePrice,
      'purchasePrice': purchasePrice,
      'stockQty': stockQty,
    };
  }

  ProductSku copyWith({
    String? id,
    String? colorName,
    String? sizeName,
    String? barcode,
    double? retailPrice,
    double? wholesalePrice,
    double? purchasePrice,
    int? stockQty,
  }) {
    return ProductSku(
      id: id ?? this.id,
      colorName: colorName ?? this.colorName,
      sizeName: sizeName ?? this.sizeName,
      barcode: barcode ?? this.barcode,
      retailPrice: retailPrice ?? this.retailPrice,
      wholesalePrice: wholesalePrice ?? this.wholesalePrice,
      purchasePrice: purchasePrice ?? this.purchasePrice,
      stockQty: stockQty ?? this.stockQty,
    );
  }
}

class Product {
  final String? id;
  final String? styleNo;
  final String? name;
  final String? brand;
  final String? season;
  final String? imageUrl;
  final String? thumbUrl;
  final List<String>? colors;
  final List<String>? sizes;
  final List<ProductSku>? skus;

  Product({
    this.id,
    this.styleNo,
    this.name,
    this.brand,
    this.season,
    this.imageUrl,
    this.thumbUrl,
    this.colors,
    this.sizes,
    this.skus,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String?,
      styleNo: json['styleNo'] as String?,
      name: json['name'] as String?,
      brand: json['brand'] as String?,
      season: json['season'] as String?,
      imageUrl: json['imageUrl'] as String?,
      thumbUrl: json['thumbUrl'] as String?,
      colors: (json['colors'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList(),
      sizes: (json['sizes'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList(),
      skus: (json['skus'] as List<dynamic>?)
          ?.map((e) => ProductSku.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'styleNo': styleNo,
      'name': name,
      'brand': brand,
      'season': season,
      'imageUrl': imageUrl,
      'thumbUrl': thumbUrl,
      'colors': colors,
      'sizes': sizes,
      'skus': skus?.map((e) => e.toJson()).toList(),
    };
  }

  Product copyWith({
    String? id,
    String? styleNo,
    String? name,
    String? brand,
    String? season,
    String? imageUrl,
    String? thumbUrl,
    List<String>? colors,
    List<String>? sizes,
    List<ProductSku>? skus,
  }) {
    return Product(
      id: id ?? this.id,
      styleNo: styleNo ?? this.styleNo,
      name: name ?? this.name,
      brand: brand ?? this.brand,
      season: season ?? this.season,
      imageUrl: imageUrl ?? this.imageUrl,
      thumbUrl: thumbUrl ?? this.thumbUrl,
      colors: colors ?? this.colors,
      sizes: sizes ?? this.sizes,
      skus: skus ?? this.skus,
    );
  }
}

File: api-spring/src/main/java/clothing/erp/entity/Product.java
package com.clothing.erp.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "product")
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shop_id", nullable = false)
    private Shop shop;

    @Column(name = "style_no", length = 100)
    private String styleNo;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(length = 100)
    private String brand;

    @Column(length = 10)
    @Enumerated(EnumType.STRING)
    private Season season;

    @Column(name = "image_url", length = 500)
    private String imageUrl;

    @Column(name = "thumb_url", length = 500)
    private String thumbUrl;

    @Column(name = "one_hand_code_preset", columnDefinition = "TEXT")
    private String oneHandCodePreset;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductColor> colors;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductSize> sizes;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProductSku> skus;

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        this.createdAt = now;
        this.updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public enum Season {
        春, 夏, 秋, 冬
    }
}
